File: blinkreceipt-demo/BlinkReceiptStaticScan/src/main/java/com/blinkreceipt/ocr/ui/MainActivity.java
package com.blinkreceipt.ocr.ui;

import android.os.Build;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import com.blinkreceipt.ocr.R;
import com.blinkreceipt.ocr.Utility;
import com.blinkreceipt.ocr.adapter.ProductsAdapter;
import com.blinkreceipt.ocr.databinding.ActivityMainBinding;
import com.blinkreceipt.ocr.presenter.MainPresenter;
import com.blinkreceipt.ocr.transfer.RecognizerResults;
import com.microblink.BlinkReceiptSdk;
import com.microblink.camera.ui.CameraCharacteristics;
import com.microblink.camera.ui.CameraRecognizerContract;
import com.microblink.camera.ui.CameraRecognizerOptions;
import com.microblink.camera.ui.CameraRecognizerResults;
import com.microblink.camera.ui.ScanCharacteristics;
import com.microblink.camera.ui.TooltipCharacteristics;
import com.microblink.core.Product;

import java.util.List;

import androidx.activity.result.ActivityResultLauncher;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.ViewGroupCompat;
import androidx.core.view.WindowCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MainActivity extends AppCompatActivity {

    private MainViewModel viewModel;

    private MainPresenter presenter;

    private ActivityMainBinding binding;

    private final ActivityResultLauncher<CameraRecognizerOptions> launcher = registerForActivityResult(new CameraRecognizerContract(), result -> {
        if (result instanceof CameraRecognizerResults.Success) {
            CameraRecognizerResults.Success results = ((CameraRecognizerResults.Success) result);

            viewModel.scanItems(new RecognizerResults(results.scanResults(), results.media()));
        } else if (result instanceof CameraRecognizerResults.Exception) {
            viewModel.scanItems(new RecognizerResults(((CameraRecognizerResults.Exception) result).exception()));
        } else {
            Toast.makeText(this, "Scan Cancelled", Toast.LENGTH_LONG).show();
        }
    });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        WindowCompat.enableEdgeToEdge(this.getWindow());

        binding = ActivityMainBinding.inflate(getLayoutInflater());
        View rootView = binding.getRoot();
        setContentView(rootView);

        ViewCompat.setOnApplyWindowInsetsListener(rootView, (v, windowInsets) -> {
            Insets insets = windowInsets.getInsets(
                    WindowInsetsCompat.Type.systemBars() |
                            WindowInsetsCompat.Type.displayCutout()
            );
            // Apply the insets as padding to the view. Here, set all the dimensions
            // as appropriate to your layout. You can also update the view's margin if
            // more appropriate.
            v.setPadding(insets.left, insets.top, insets.right, insets.bottom);

            // Return CONSUMED if you don't want the window insets to keep passing down
            // to descendant views.
            return WindowInsetsCompat.CONSUMED;
        });
        ViewGroupCompat.installCompatInsetsDispatch(rootView);
        WindowCompat.getInsetsController(getWindow(), getWindow().getDecorView())
                .setAppearanceLightStatusBars(true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            this.getWindow().setStatusBarContrastEnforced(true);
        }

        viewModel = new ViewModelProvider(this).get(MainViewModel.class);

        presenter = new MainPresenter();

        final RecyclerView recyclerView = findViewById(R.id.products);

        final ProductsAdapter adapter = new ProductsAdapter();

        LinearLayoutManager manager = new LinearLayoutManager(this);

        recyclerView.addItemDecoration(new DividerItemDecoration(this, manager.getOrientation()));

        recyclerView.setLayoutManager(manager);

        recyclerView.setAdapter(adapter);

        viewModel.scanItems().observe(this, results -> {
            if (results != null) {
                if (presenter.exception(results)) {
                    Throwable e = results.e();

                    Toast.makeText(MainActivity.this, e != null ? e.toString() :
                            getString(R.string.no_products_found_on_receipt), Toast.LENGTH_LONG).show();

                    return;
                }

                List<Product> products = presenter.products(results);

                if (Utility.isNullOrEmpty(products)) {
                    Toast.makeText(MainActivity.this, R.string.no_products_found_on_receipt, Toast.LENGTH_SHORT).show();

                    return;
                }

                adapter.addAll(products);
            } else {
                Toast.makeText(MainActivity.this, R.string.no_products_found_on_receipt, Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();

        inflater.inflate(R.menu.main_menu, menu);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.sdk_version) {
            new AlertDialog.Builder(this)
                    .setTitle(R.string.sdk_version_dialog_title)
                    .setMessage(BlinkReceiptSdk.versionName(this))
                    .setPositiveButton(android.R.string.ok, (dialog, which) -> dialog.dismiss())
                    .create()
                    .show();

            return true;
        }

        if (item.getItemId() == R.id.camera) {
            launcher.launch(new CameraRecognizerOptions.Builder()
                    .options(viewModel.scanOptions())
                    .characteristics(new CameraCharacteristics.Builder()
                            .cameraPermission(true)
                            .scanCharacteristics( new ScanCharacteristics.Builder()
                                    .total(true)
                                    .merchant(true)
                                    .date(true)
                                    .build())
                            .tooltipCharacteristics(
                                    new TooltipCharacteristics.Builder()
                                            .displayTooltips(true)
                                            .build())
                            .build())
                    .build());

            return true;
        }

        return super.onOptionsItemSelected(item);
    }

}

File: blinkreceipt-demo/BlinkReceiptScanPdf/src/main/java/com/blinkreceipt/scan/pdf/MainActivity.kt
package com.blinkreceipt.scan.pdf

import android.content.ContentResolver
import android.net.Uri
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.ParcelFileDescriptor
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.VisibleForTesting
import androidx.core.view.ViewCompat
import androidx.core.view.ViewGroupCompat
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.blinkreceipt.scan.pdf.databinding.ActivityMainBinding
import com.blinkreceipt.scan.pdf.internal.PdfViewer
import com.blinkreceipt.scan.pdf.internal.PdfViewerAdapter
import com.microblink.PdfClient
import com.microblink.ScanOptions
import com.microblink.internal.READ_MODE

class MainActivity : AppCompatActivity() {

    private var _binding: ActivityMainBinding? = null

    @VisibleForTesting
    internal val binding
        get() = _binding!!

    private lateinit var client: PdfClient

    private var document: Uri? = null

    private val pdfViewer: PdfViewer by lazy {
        PdfViewer()
    }

    private lateinit var pdfAdapter: PdfViewerAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.enableEdgeToEdge(this.window)

        _binding = ActivityMainBinding.inflate(layoutInflater)
        val rootView: View = binding.getRoot()
        setContentView(rootView)

        ViewCompat.setOnApplyWindowInsetsListener(
            rootView
        ) { v: View, windowInsets: WindowInsetsCompat? ->
            val insets = windowInsets!!.getInsets(
                WindowInsetsCompat.Type.systemBars() or
                        WindowInsetsCompat.Type.displayCutout()
            )
            // Apply the insets as padding to the view. Here, set all the dimensions
            // as appropriate to your layout. You can also update the view's margin if
            // more appropriate.
            v.setPadding(insets.left, insets.top, insets.right, insets.bottom)
            WindowInsetsCompat.CONSUMED
        }
        ViewGroupCompat.installCompatInsetsDispatch(rootView)

        WindowCompat.getInsetsController(
            window,
            window.decorView
        ).isAppearanceLightStatusBars = true
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            @Suppress("DEPRECATION")
            this.window.setStatusBarContrastEnforced(true)
        }

        client = PdfClient(applicationContext).also {
            val options: ScanOptions = ScanOptions.newBuilder().build()

            it.countryCode = options.countryCode()

            it.filterSensitiveData = options.filterSensitiveData()

            it.returnSubProducts = options.returnSubProducts()

            it.returnVoidedProducts = options.returnVoidedProducts()

            lifecycle.addObserver(it)
        }

        pdfAdapter = PdfViewerAdapter(this, pdfViewer)

        lifecycle.addObserver(pdfViewer)

        val launcher = registerForActivityResult(ActivityResultContracts.OpenDocument()) { uri ->
            document = uri

            uri?.let {
                contentResolver.toFileDescriptor(uri)?.let { contents ->
                    runCatching {
                        pdfViewer.parse(contents).also {
                            pdfAdapter.notifyItemRangeInserted(0, pdfViewer.size())
                        }

                        contents.close()
                    }.onFailure {
                        Toast.makeText(applicationContext, it.toString(), Toast.LENGTH_LONG).show()
                    }
                }
            }
        }

        binding.selectPdf.setOnClickListener {
            pdfAdapter.clear()

            launcher.launch(arrayOf(MIME_TYPE))
        }

        binding.scanPdf.setOnClickListener {
            document?.let { uri ->
                Toast.makeText(applicationContext, "Scanning PDF", Toast.LENGTH_SHORT).show()

                runCatching {
                    client.recognize(uri)
                        .addOnSuccessListener(this) { results ->
                            Log.d(TAG, results.toString())

                            Toast.makeText(
                                applicationContext,
                                results.toString(),
                                Toast.LENGTH_LONG
                            )
                                .show()
                        }.addOnFailureListener {
                            Toast.makeText(applicationContext, it.toString(), Toast.LENGTH_LONG)
                                .show()
                        }
                }.onFailure {
                    Toast.makeText(applicationContext, it.toString(), Toast.LENGTH_LONG).show()
                }
            } ?: Toast.makeText(
                applicationContext,
                getString(R.string.please_select_a_pdf), Toast.LENGTH_SHORT,
            ).show()
        }

        with(binding.pdfViewer) {
            layoutManager = LinearLayoutManager(this@MainActivity).apply {
                orientation = LinearLayoutManager.HORIZONTAL
            }
            adapter = pdfAdapter
        }
    }

    private fun ContentResolver.toFileDescriptor(uri: Uri): ParcelFileDescriptor? =
        openFileDescriptor(uri, READ_MODE)


    internal companion object {

        const val MIME_TYPE = "application/pdf"

        const val TAG = "PdfMainActivity"
    }
}